File: 39.combination-sum.java
/*
 * @lc app=leetcode id=39 lang=java
 *
 * [39] Combination Sum
 */

// @lc code=start
class Solution {

    List<List<Integer>> res = new LinkedList<>();

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        if (candidates.length == 0) return res;
        Arrays.sort(candidates);
        backtrack(0, candidates, target, new ArrayList<>());
        return res;
    }

    private void backtrack(int start, int[] candidates, int target, List<Integer> tmpList) {
        if (target == 0) {
            res.add(new ArrayList<>(tmpList));
            return;
        }
        for (int i = start; i < candidates.length && candidates[i] <= target; ++i) {
            tmpList.add(candidates[i]);
            backtrack(i, candidates, target-candidates[i], tmpList);
            tmpList.remove(tmpList.size()-1);
        }
    }
}
// @lc code=end

File: go/39.combination-sum.go
/*
 * @lc app=leetcode id=39 lang=golang
 *
 * [39] Combination Sum
 */

// @lc code=start
func combinationSum(candidates []int, target int) [][]int {
	sort.Ints(candidates)
	res := [][]int{}
	var backtrack func(int, int, []int)
	backtrack = func(start, acc int, comb []int) {
		if acc == target {
			tmp := make([]int, len(comb))
			copy(tmp, comb)
			res = append(res, tmp)
			return
		}
		for i := start; i < len(candidates) && candidates[i] + acc <= target; i++ {
			cnt, acc2 := 1, candidates[i] + acc
			for acc2 <= target {
				comb = append(comb, candidates[i])
				backtrack(i+1, acc2, comb)
				cnt, acc2 = cnt+1, candidates[i] + acc2
			}
			comb = comb[:len(comb)-cnt+1]
		}
	}
	backtrack(0, 0, []int{})
	return res
}
// @lc code=end

